libyul/optimiser/UnusedFunctionArgumentPruner.cpp

```diff
@@ -0,0 +1,191 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * UnusedFunctionArgumentPruner: Optimiser step that removes unused-parameters function.
+ */
+
+#include <libyul/optimiser/UnusedFunctionArgumentPruner.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/NameDisplacer.h>
+#include <libyul/AsmData.h>
+
+#include <libsolutil/CommonData.h>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+using namespace solidity::langutil;
+
+namespace
+{
+
+/**
+ * First step of UnusedFunctionArgumentPruner: Find functions with whose parameters are not used in
+ * its body.
+ */
+struct FindFunctionsWithUnusedParameters
+{
+	void operator()(FunctionDefinition const& _function)
+	{
+		auto namesFound = ReferencesCounter::countReferences(_function.body);
+
+		TypedNameList reducedParameters;
+
+		for (auto const& parameter: _function.parameters)
+			if (namesFound.count(parameter.name))
+				reducedParameters.push_back(parameter);
+
+		if (reducedParameters.size() < _function.parameters.size())
+		{
+			functions.insert(_function.name);
+			prunedTypeNames[_function.name] = move(reducedParameters);
+		}
+	}
+
+	set<YulString> functions;
+	map<YulString, TypedNameList> prunedTypeNames;
+};
+
+/**
+ * Second step of UnusedFunctionArgumentPruner: replace all references to functions with unused
+ * parameters with a new name.
+ *
+ * For example: `function f(x) -> y { y := 1 }` will be replaced with something
+ * like : `function f_1(x) -> y { y := 1 }`  and all references to `f` by `f_1`.
+ */
+struct ReplaceFunctionName: public NameDisplacer
+{
+	explicit ReplaceFunctionName(
+		NameDispenser& _dispenser,
+		std::set<YulString> const& _namesToFree
+	): NameDisplacer(_dispenser, _namesToFree) {}
+
+	std::map<YulString, YulString>& translations() { return m_translations; }
```

Comment:
Maybe it would be better to rather add this function to NameDisplacer. If we ever modify the behaviour of NameDisplacer we will better notice.